### PR TITLE
[GlobalOpt] Handle DL.getAllocaAddrSpace() != 0

### LINKED ISSUE #65155: [Clang][OpenMP] Compilation Aborted (core dumped)
Getting an Aborted (core dumped) from compiling with clang. You can reproduce the core dump with the attached reduced.ll file using ```opt -p globalopt reduced.ll```

reduced.ll:
```
; ModuleID = 'reduced.bc'
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8"
target triple = "amdgcn-amd-amdhsa"

@x = internal global i32 poison

; Function Attrs: norecurse
define amdgpu_kernel void @__omp_offloading_10306_92b48f5_test_uses_allocators_large_cap_l31(i32 %0) #0 {
entry:
  store i32 %0, ptr @x, align 4
  ret void
}

attributes #0 = { norecurse }
```

### PR DESCRIPTION
**Pull request by tylanphear**
Fix crash on RAUW due to locals and globals having different address
spaces. This is the intent of the original code, but it assumes the
alloca address space is 0. This patch fixes the code to check that the
global's address space matches `DL.getAllocaAddrSpace()` instead.

Fixes #65155

### CODE REVIEW

File: llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8"
+target triple = "amdgcn-amd-amdhsa"
+
+; Check that we don't convert the global into an alloca if their respective address
+; spaces differ, and the alloca addrspace is non-zero.
+
+@x = internal global i32 poison
+
+; Function Attrs: norecurse
+define amdgpu_kernel void @test(i32 %0) #0 {
```
Comment (arsenm) [inline llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll:12]: Don't need amdgpu_kernel

File: llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll [hunk 2]
```diff
@@ -0,0 +1,23 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt -S -passes=globalopt < %s | FileCheck %s
+target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8"
+target triple = "amdgcn-amd-amdhsa"
```
Comment (arsenm) [inline llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll:4]: Don't need the triple, this would unnecessarily require the target to be built

**Review by arsenm — CHANGES_REQUESTED** [review]
LGTM except the test needs cleanup to avoid failing when AMDGPU isn't built

File: llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll (hunk 1 above)
Comment (tylanphear) [reply llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll]: Removed.

File: llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll (hunk 2 above)
Comment (tylanphear) [reply llvm/test/Transforms/GlobalOpt/global_alloca_diff_addrspaces.ll]: Removed.

**Review by arsenm — APPROVED** [review]

**Pull request merged**